<a href="https://colab.research.google.com/github/el-grudge/mleng-zoomcamp/blob/main/week_8/homework_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!wget https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip; unzip -q data.zip

--2023-11-20 00:55:17--  https://github.com/SVizor42/ML_Zoomcamp/releases/download/bee-wasp-data/data.zip
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/e6c56cb7-dce1-463f-865b-01e913c38485?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20231120%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20231120T005517Z&X-Amz-Expires=300&X-Amz-Signature=9c92ab3ec8e7e76636cd10290e17e88aaab32df38ec3cdffae070b6c6ef00ec3&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=405934815&response-content-disposition=attachment%3B%20filename%3Ddata.zip&response-content-type=application%2Foctet-stream [following]
--2023-11-20 00:55:17--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/405934815/e6c56cb7-dce1-463f-865b-01e913c38485?X-Amz-A

In [2]:
import numpy as np
import matplotlib.pyplot as plt

from tensorflow.keras import layers
from tensorflow.keras import models
from tensorflow.keras import optimizers
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.preprocessing import image

In [3]:
# model parameters
filter_count = 32
filter_size = (3,3)
input_shape = (150,150,3)
pooling_size = (2,2)

In [4]:
# model definition
model = models.Sequential()
model.add(layers.Conv2D(filter_count, filter_size, activation='relu', input_shape=input_shape))
model.add(layers.MaxPooling2D(pooling_size))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [5]:
# compile model
model.compile(loss='binary_crossentropy',
             optimizer=optimizers.SGD(learning_rate=0.002, momentum=0.8),
             metrics=['acc'])

#### Question 1

Since we have a binary classification problem, what is the best loss function for us?

* `mean squared error`
* `binary crossentropy`    
* `categorical crossentropy`    
* `cosine similarity`

Answer: `binary crossentropy`

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 74, 74, 32)        0         
 D)                                                              
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11215873 (42.79 MB)
Trainable params: 11215873 (42.79 MB)
Non-trainable params: 0 (0.00 Byte)
______________

#### Question 2

What's the number of parameters in the convolutional layer of our model? You can use the summary method for that.

Answer: 11214912

In [7]:
# define data generators
train_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/train',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 3677 images belonging to 2 classes.


In [8]:
val_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/test',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 918 images belonging to 2 classes.


In [9]:
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=val_generator
)

Epoch 1/10
184/184 [==============================] - 19s 40ms/step - loss: 0.6721 - acc: 0.5747 - val_loss: 0.7491 - val_acc: 0.5370
Epoch 2/10
184/184 [==============================] - 8s 42ms/step - loss: 0.6171 - acc: 0.6557 - val_loss: 0.5733 - val_acc: 0.7092
Epoch 3/10
184/184 [==============================] - 9s 47ms/step - loss: 0.5704 - acc: 0.7055 - val_loss: 0.5401 - val_acc: 0.7244
Epoch 4/10
184/184 [==============================] - 7s 39ms/step - loss: 0.5308 - acc: 0.7425 - val_loss: 0.5388 - val_acc: 0.7168
Epoch 5/10
184/184 [==============================] - 9s 48ms/step - loss: 0.5136 - acc: 0.7520 - val_loss: 0.5214 - val_acc: 0.7669
Epoch 6/10
184/184 [==============================] - 9s 46ms/step - loss: 0.4938 - acc: 0.7773 - val_loss: 0.5348 - val_acc: 0.7298
Epoch 7/10
184/184 [==============================] - 7s 39ms/step - loss: 0.4830 - acc: 0.7813 - val_loss: 0.5340 - val_acc: 0.7353
Epoch 8/10
184/184 [==============================] - 9s 48ms/step -

#### Question 3

What is the median of training accuracy for all the epochs for this model?

* 0.20
* 0.40
* 0.60
* 0.80

Answer: 0.80

In [10]:
np.median(history.history['acc'])

0.7646178901195526

Question 4

What is the standard deviation of training loss for all the epochs for this model?

* 0.031
* 0.061
* 0.091
* 0.131

Answer: 0.091

In [11]:
np.std(history.history['loss'])

0.08097588912086935

In [12]:
train_generator = ImageDataGenerator(
    rescale=1./255,
    rotation_range=50,
    width_shift_range=0.1,
    height_shift_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest').flow_from_directory(
        './data/train',
        target_size=(150,150),
        batch_size=20,
        class_mode='binary',
        shuffle=True)

Found 3677 images belonging to 2 classes.


In [13]:
val_generator = ImageDataGenerator(rescale=1./255).flow_from_directory(
    './data/test',
    target_size=(150,150),
    batch_size=20,
    class_mode='binary',
    shuffle=True
)

Found 918 images belonging to 2 classes.


In [14]:
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=val_generator
)

Epoch 1/10
184/184 [==============================] - 28s 152ms/step - loss: 0.4957 - acc: 0.7786 - val_loss: 0.5398 - val_acc: 0.7538
Epoch 2/10
184/184 [==============================] - 28s 151ms/step - loss: 0.5060 - acc: 0.7648 - val_loss: 0.4722 - val_acc: 0.7778
Epoch 3/10
184/184 [==============================] - 28s 152ms/step - loss: 0.4779 - acc: 0.7830 - val_loss: 0.4854 - val_acc: 0.7767
Epoch 4/10
184/184 [==============================] - 27s 145ms/step - loss: 0.4855 - acc: 0.7696 - val_loss: 0.4903 - val_acc: 0.7734
Epoch 5/10
184/184 [==============================] - 28s 151ms/step - loss: 0.4664 - acc: 0.7803 - val_loss: 0.4729 - val_acc: 0.7712
Epoch 6/10
184/184 [==============================] - 28s 152ms/step - loss: 0.4728 - acc: 0.7800 - val_loss: 0.5039 - val_acc: 0.7745
Epoch 7/10
184/184 [==============================] - 28s 154ms/step - loss: 0.4672 - acc: 0.7841 - val_loss: 0.4972 - val_acc: 0.7680
Epoch 8/10
184/184 [==============================] - 2

Question 5

Let's train our model for 10 more epochs using the same code as previously.

    Note: make sure you don't re-create the model - we want to continue training the model we already started training.

What is the mean of test loss for all the epochs for the model trained with augmentations?


* 0.18
* 0.48
* 0.78
* 0.108

Answer: 0.48

In [15]:
np.mean(history.history['val_loss'])

0.4897406905889511

Question 6

What's the average of test accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

* 0.38
* 0.58
* 0.78
* 0.98

Answer: 0.78

In [16]:
np.mean(history.history['val_acc'][5:10])

0.7760348439216613